src/distributions/uniform.rs

```diff
@@ -714,20 +714,30 @@ impl SampleUniform for char {
 ///
 /// Unless you are implementing [`UniformSampler`] for your own type, this type
 /// should not be used directly, use [`Uniform`] instead.
+///
+/// This differs from integer range sampling since the range `0xD800..=0xDFFF`
+/// are used for surrogate pairs in UCS and UTF-16, and consequently are not
+/// valid Unicode code points. We must therefore avoid sampling values in this
+/// range.
 #[derive(Clone, Copy, Debug)]
 #[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
 pub struct UniformChar {
     sampler: UniformInt<u32>,
 }
 
+/// UTF-16 surrogate range start
+const CHAR_SURROGATE_START: u32 = 0xD800
+/// UTF-16 surrogate range size
+const CHAR_SURROGATE_LEN: u32 = 0xE000 - CHAR_SURROGATE_START;
+
 impl UniformChar {
     #[inline]
     fn new_(mut low: u32, mut high: u32) -> Self {
-        if low >= 0xD800 {
-            low -= 0xE000 - 0xD800;
+        if low >= CHAR_SURROGATE_START {
+            low -= CHAR_SURROGATE_LEN;
```

Comment:
I don't think I understand this. Shouldn't it be `low = CHAR_SURROGATE_START - 1`?

Reply:
No. Rather than switching in the sampler whether we need to add the gap, we just add it for all values over the start (which means that low must be reduced by the gap size if it's after the gap).